File: packages/shared/src/types/meters.ts
export interface Meter {
  id: string;
  room_id: string;
  service_id: string;
  service_name?: string | null;
  service_unit?: string | null;
  serial_number: string | null;
  initial_reading: string; // Decimal as string
  latest_reading?: string | null; // Decimal as string
  latest_reading_date?: string | null; // ISO Date String
  status: 'ACTIVE' | 'INACTIVE';
  created_at: string;
  updated_at: string;
}

export type MeterResponse = Meter;

export interface MeterReading {
  id: string;
  meter_id: string;
  reading_value: string; // Decimal as string
  reading_date: string;
  image_url: string | null;
  usage: string; // Decimal as string
  tenant_id: string | null;
  period_month: string | null;
  created_at: string;
}

export type MeterReadingResponse = MeterReading;

export interface CreateMeterInput {
  room_id: string;
  service_id: string;
  serial_number?: string | null;
  initial_reading?: string;
}

export interface RecordReadingParams {
  reading_value: string;
  reading_date?: string;
  image_url?: string | null;
  period_month?: string | null;
}

export type RecordReadingInput = RecordReadingParams;

export interface LandlordMeterSummary {
  total_meters: number;
  pending_readings: number;
  overdue_readings: number;
  submission_rate: number;
}

export interface LandlordMeterDetail {
  id: string;
  room_id: string;
  room_code: string;
  building_id: string;
  building_name: string;
  service_name: string;
  service_unit: string;
  serial_number: string | null;
  status: 'SUBMITTED' | 'PENDING' | 'OVERDUE';
  last_reading: string | null;
  last_reading_date: string | null;
}

File: apps/api/server-api/src/views/meters.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;
use rust_decimal::Decimal;
use chrono::{DateTime, Utc};
use crate::models::_entities::{meters, services};

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct CreateMeterParams {
    pub room_id: Uuid,
    pub service_id: Uuid,
    pub serial_number: Option<String>,
    pub initial_reading: Option<Decimal>,
}



#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct MeterResponse {
    pub id: Uuid,
    pub room_id: Uuid,
    pub service_id: Uuid,
    pub service_name: Option<String>,
    pub service_unit: Option<String>,
    pub serial_number: Option<String>,
    pub initial_reading: Decimal,
    pub latest_reading: Option<Decimal>,
    pub latest_reading_date: Option<DateTime<Utc>>,
    pub latest_reading_period: Option<String>,
    pub latest_reading_status: Option<String>,
    pub latest_reading_error: Option<String>,
    pub latest_usage: Option<Decimal>,
    pub status: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}



impl MeterResponse {
    pub fn from_model(m: meters::Model, s: Option<services::Model>) -> Self {
        Self {
            id: m.id,
            room_id: m.room_id,
            service_id: m.service_id,
            service_name: s.as_ref().map(|x| x.name.clone()),
            service_unit: s.as_ref().map(|x| x.unit.clone()),
            serial_number: m.serial_number,
            initial_reading: m.initial_reading,
            latest_reading: None,
            latest_reading_date: None,
            latest_reading_period: None,
            latest_reading_status: None,
            latest_reading_error: None,
            latest_usage: None,
            status: m.status,
            created_at: m.created_at.into(),
            updated_at: m.updated_at.into(),
        }
    }
}

impl From<meters::Model> for MeterResponse {
    fn from(m: meters::Model) -> Self {
        Self {
            id: m.id,
            room_id: m.room_id,
            service_id: m.service_id,
            service_name: None,
            service_unit: None,
            serial_number: m.serial_number,
            initial_reading: m.initial_reading,
            latest_reading: None,
            latest_reading_date: None,
            latest_reading_period: None,
            latest_reading_status: None,
            latest_reading_error: None,
            latest_usage: None,
            status: m.status,
            created_at: m.created_at.into(),
            updated_at: m.updated_at.into(),
        }
    }
}



#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct LandlordMeterSummary {
    pub total_meters: u64,
    pub pending_readings: u64,
    pub overdue_readings: u64,
    pub submission_rate: f32, // Percentage
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct LandlordMeterDetail {
    pub id: Uuid,
    pub room_id: Uuid,
    pub room_code: String,
    pub building_id: Uuid,
    pub building_name: String,
    pub service_name: String,
    pub service_unit: String,
    pub serial_number: Option<String>,
    pub status: String, // SUBMITTED, PENDING, OVERDUE
    pub last_reading: Option<Decimal>,
    pub last_reading_date: Option<DateTime<Utc>>,
}
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct LandlordListParams {
    pub building_id: Option<Uuid>,
    pub period_month: Option<String>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct UpdateMeterStatusParams {
    pub status: String,
}


